### PR TITLE
feat: Send pings & receive pongs

### PR DESCRIPTION
**Pull request by sdsantos**

### CODE REVIEW

File: app/src/main/java/tech/relaycorp/ping/domain/SendPing.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    suspend fun send(peer: Peer, duration: Duration): String {
+        val senderAddress = appPreferences.firstPartyEndpointAddress().first()
+            ?: throw SendPingException("Sender not set")
+        val sender = firstPartyEndpointLoad.load(senderAddress)
+            ?: throw SendPingException("Sender not registered")
+
+        val recipient = publicThirdPartyEndpointLoad.load(peer.privateAddress)
+            ?: throw SendPingException("Recipient not imported")
+
+        val parcelId = ParcelId.generate()
```
Comment (gnarea) [inline app/src/main/java/tech/relaycorp/ping/domain/SendPing.kt:36]: I think we shouldn't be using the `ParcelId` here: The parcel id and the ping id are completely independent of each other -- they just happen to be UUID4 values.

I'd suggest to let awaladroid generate a random parcel id for the message, and set the ping id here by calling something like `UUID.randomUUID().toString()`.

Thoughts?

File: app/src/main/res/xml/network_security_config.xml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<network-security-config>
+    <!-- Allow local cleartext networking with the PDC Server -->
+    <domain-config cleartextTrafficPermitted="true">
+        <domain includeSubdomains="false">127.0.0.1</domain>
```
Comment (gnarea) [inline app/src/main/res/xml/network_security_config.xml:5]: Can this go away with the latest version of awaladroid?

**Review by gnarea — CHANGES_REQUESTED** [review]
On the ping screen, I think it'd be good to show something when the ping was sent but the pong hasn't been received yet. Especially when the user is using the app for the first time and the pong takes a while to arrive. For example, we could display the `Pong received at` field with the value `Pending`. Thoughts?

---

On the send ping screen, the app crashes if I change the unit when the value isn't valid in the new unit. For example, if the original value is "25 days" and I change the unit to "months", the following happens:

```
    --------- beginning of crash
2021-04-02 18:54:21.023 9303-9303/tech.relaycorp.ping E/AndroidRuntime: FATAL EXCEPTION: main
    Process: tech.relaycorp.ping, PID: 9303
    java.lang.ArrayIndexOutOfBoundsException: length=6; index=8
        at java.util.Arrays$ArrayList.get(Arrays.java:3769)
        at tech.relaycorp.ping.ui.ping.SendPingActivity$setupExpiresAt$1.invoke(SendPingActivity.kt:118)
        at tech.relaycorp.ping.ui.ping.SendPingActivity$setupExpiresAt$1.invoke(SendPingActivity.kt:24)
        at tech.relaycorp.ping.ui.common.SimpleItemSelectedListener.onItemSelected(SimpleItemSelectedListener.kt:11)
        at android.widget.AdapterView.fireOnSelected(AdapterView.java:944)
        at android.widget.AdapterView.dispatchOnItemSelected(AdapterView.java:933)
        at android.widget.AdapterView.-wrap1(Unknown Source:0)
        at android.widget.AdapterView$SelectionNotifier.run(AdapterView.java:898)
        at android.os.Handler.handleCallback(Handler.java:790)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```

**Comment by gnarea** [comment]
I've been playing a bit more with this and I've found two small issues when sending a ping:

- The time in `Pong received at` often changes as soon as the pong is received. For example, it it's received at `19:07:27`, it'd quickly change to `19:07:28`.
- When I press send, it takes a while to get the ping screen. Like 1-2 seconds. Presumably because of all the heavy lifting done by awaladroid in the background. Are we doing this in a different thread, and if not, should we?

**Comment by gnarea** [comment]
> On the ping screen, I think it'd be good to show something when the ping was sent but the pong hasn't been received yet. Especially when the user is using the app for the first time and the pong takes a while to arrive. For example, we could display the `Pong received at` field with the value `Pending`. Thoughts?

This is technically unrelated to this PR so I moved it to https://github.com/relaycorp/awala-ping-android/issues/27

**Comment by sdsantos** [comment]
> I've been playing a bit more with this and I've found two small issues when sending a ping:
> 
> * The time in `Pong received at` often changes as soon as the pong is received. For example, it it's received at `19:07:27`, it'd quickly change to `19:07:28`.
> * When I press send, it takes a while to get the ping screen. Like 1-2 seconds. Presumably because of all the heavy lifting done by awaladroid in the background. Are we doing this in a different thread, and if not, should we?

- The time change is weird. Is the server sending double-pongs (not filtering that out, and we're accepting the new timestamp), or did you send multiple pings with multiple taps on the send icon?

- It's on a background thread, but we are not giving any clue that stuff is happening. I'll add a loader and disable the button.

**Comment by gnarea** [comment]
> The time change is weird. Is the server sending double-pongs (not filtering that out, and we're accepting the new timestamp), or did you send multiple pings with multiple taps on the send icon?

No, I can see in the Android gateway logs that there's only one coming in.

> It's on a background thread, but we are not giving any clue that stuff is happening. I'll add a loader and disable the button.

:+1:

File: app/src/main/java/tech/relaycorp/ping/domain/ReceivePong.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+class ReceivePong
+@Inject constructor(
+    private val pingDao: PingDao,
+    private val pingSerialization: PingSerialization
+) {
+
+    suspend fun receive(incomingMessage: IncomingMessage) {
+        if (incomingMessage.type != AwalaPing.V1.PongType) {
+            incomingMessage.ack()
```
Comment (gnarea) [inline app/src/main/java/tech/relaycorp/ping/domain/ReceivePong.kt:21]: This should also return to skip the rest of the function

File: app/src/main/java/tech/relaycorp/ping/domain/ReceivePong.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        val pingId = pingSerialization.extractPingIdFromPong(
+            incomingMessage.content
+        )
+        val ping = pingDao.getPublic(pingId).first()?.ping ?: run {
+            logger.log(Level.INFO, "Received pong for unknown ping $pingId")
```
Comment (gnarea) [inline app/src/main/java/tech/relaycorp/ping/domain/ReceivePong.kt:28]: It should also be ack'ed here

File: app/src/main/res/layout/view_loading.xml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        android:indeterminate="true" />
+
+    <TextView
+        android:id="@+id/loadingMessage"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:gravity="center_horizontal"
+        android:textColor="?colorOnSurface"
+        android:textIsSelectable="true"
+        tools:text="Loading..." />
```
Comment (gnarea) [inline app/src/main/res/layout/view_loading.xml:25]: I think this should be translatable

**Review by gnarea — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/view_loading.xml (hunk 5 above)
Comment (sdsantos) [reply app/src/main/res/layout/view_loading.xml:25]: That's under `tools`, it's just for displaying on Android Studio, it does not end up on the release APK. The actual text is set whenever we use it.

Comment (gnarea) [reply app/src/main/res/layout/view_loading.xml:25]: Gotcha!

File: app/src/main/java/tech/relaycorp/ping/domain/ReceivePong.kt (hunk 3 above)
Comment (sdsantos) [reply app/src/main/java/tech/relaycorp/ping/domain/ReceivePong.kt:21]: Silly me.

**Review by gnarea — APPROVED** [review]

**Pull request merged**